federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncService(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing service: %v", err)
+			}
+		}()
+	}
+}
+
+func (s *ServiceController) init() error {
+//TODO: waiting for DNS lib
+
+	zones, ok := s.dns.Zones()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support zone enumeration, which is required for creating load balancers.")
+	}
+	zoneList, err := zones.List()
+	if err != nil {
+		return fmt.Errorf("failed to get zone from cloud provider, will not be able to create load balancers: %v", err)
+	}
+	s.zones = zoneList
+	return nil
+}
+
+func wantsDNSRecords(service *v1.Service) bool{
+	return true
+}
+
+func (s *ServiceController) processServiceForCluster(key, clusterName string, service, appliedState *v1.Service, client clientset.Clientset) (error, bool) {
+	// Create or Update k8s Service
+	s.ensureClusterServices(service, client, clusterName)
+	if appliedState != nil && !s.needsUpdate(appliedState, service) {
+		glog.Infof("DNS records don't need update for service %s", key)
+		return nil, notRetryable
+	}
+
+	// Note: It is safe to just call EnsureDNSRecords.  But, on some clouds that requires a delete & create,
+	// which may involve service interruption.  Also, we would like user-friendly events.
+
+	// Save the state so we can avoid a write if it doesn't change
+	//TODO: jesse
+	//previousState := api.LoadBalancerStatusDeepCopy(&service.Status.LoadBalancer) // TODO: quinton: are the DNS names in here?  I think so.
+	if !wantsDNSRecords(service) {
+		needDelete := true
+		if appliedState != nil {
+			if !wantsDNSRecords(appliedState) {
+				needDelete = false
+			}
+		} else {
+			// If we don't have any cached memory of the DNS records, we have to ask
+			// the cloud provider for what it knows about them.
+			// Technically EnsureDNSRecordsDeleted can cope, but we want to post meaningful events
+			//TODO: Jesse:uncomment
+			//_, exists, err := s.dns.GetDNSRecords(service)
+			//if err != nil {
+			//	return fmt.Errorf("Error getting DNS records for service %s: %v", namespacedName, err), retryable
+			//}
+			//if !exists {
+			//	needDelete = false
+			//}
+		}
+
+		if needDelete {
+			glog.Infof("Deleting existing DNS records for service %s that no longer needs them.", key)
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletingDNSRecords", "Deleting DNS records")
+			//todo:Jesse
+			//if err := s.dns.EnsureDNSRecordsDeleted(service); err != nil {
+			//	return err, retryable
+			//}
+			s.eventRecorder.Event(service, api.EventTypeNormal, "DeletedDNSRecords", "Deleted DNS records")
+		}
+
+		service.Status.LoadBalancer = v1.LoadBalancerStatus{} // TODO: quinton: is the DNS name in here?
```

Comment:
See above comments about load balancer status reporting.  I assume that service here is the Ubernetes service, not the Kubernetes service?